File: JAVA/Cocktail Sort/StackPointerCocktailSort.java
import java.io.*;

public class StackPointerCocktailSort
{
    static class Rating
    {
        int userId, movieId;
        float rating;
        long timestamp;

        Rating(int u, int m, float r, long t)
        {
            userId = u; movieId = m; rating = r; timestamp = t;
        }

        int compareTo(Rating o)
        {
            return Float.compare(this.rating, o.rating);
        }

        public String toString()
        {
            return userId + "," + movieId + "," + rating + "," + timestamp;
        }
    }

    static class Node
    {
        Rating r;
        Node prev, next;

        Node(Rating r) { this.r = r; }
    }

    Node top = null;
    int size = 0;

    void push(Rating r)
    {
        Node n = new Node(r);
        if (top != null)
        {
            top.next = n;
            n.prev = top;
        }
        top = n;
        size++;
    }

    void cocktailSort()
    {
        if (size <= 1) return;

        Node base = top;
        while (base.prev != null) base = base.prev;

        boolean swapped = true;
        Node start = base, end = top;

        while (swapped)
        {
            swapped = false;
            for (Node c = start; c != end; c = c.next)
            {
                if (c.r.compareTo(c.next.r) > 0)
                {
                    Rating tmp = c.r;
                    c.r = c.next.r;
                    c.next.r = tmp;
                    swapped = true;
                }
            }

            if (!swapped) break;
            swapped = false;
            end = end.prev;

            for (Node c = end; c != start; c = c.prev)
            {
                if (c.prev.r.compareTo(c.r) > 0)
                {
                    Rating tmp = c.r;
                    c.r = c.prev.r;
                    c.prev.r = tmp;
                    swapped = true;
                }
            }
            start = start.next;
        }
    }

    public static void main(String[] args)
    throws Exception
    {
        if (args.length < 2)
        {
            System.out.println("Uso: java StackPointerCocktailSort <ratings.csv> <N>");
            return;
        }

        String path = args[0];
        int N = Integer.parseInt(args[1]);
        StackPointerCocktailSort pilha = new StackPointerCocktailSort();

        try (BufferedReader br = new BufferedReader(new FileReader(path)))
        {
            String line = br.readLine();
            int count = 0;
            while ((line = br.readLine()) != null && count < N)
            {
                String[] f = line.split(",");
                pilha.push(new Rating(Integer.parseInt(f[0]), Integer.parseInt(f[1]),
                                      Float.parseFloat(f[2]), Long.parseLong(f[3])));
                count++;
            }
        }

        long t0 = System.nanoTime();
        pilha.cocktailSort();
        long t1 = System.nanoTime();

        System.out.printf("Ordenado %d registros em %.2f ms\n", N, (t1 - t0) / 1e6);

        Node cur = pilha.top;
        while (cur.prev != null) cur = cur.prev;

        for (int i = 0; i < 100 && cur != null; i++, cur = cur.next)
        {
            System.out.println(cur.r);
        }
    }
}

File: Ruby/CocktailSort/stack_pointer_cocktail_sort.rb
require 'benchmark'
require 'objspace'

class Rating
  attr_accessor :user_id, :movie_id, :rating, :timestamp
  def initialize(user_id, movie_id, rating, timestamp)
    @user_id = user_id
    @movie_id = movie_id
    @rating = rating
    @timestamp = timestamp
  end
  def <=>(other)
    @rating <=> other.rating
  end
  def to_s
    "#{@user_id},#{@movie_id},#{@rating},#{@timestamp}"
  end
end

class Node
  attr_accessor :rating, :next, :prev
  def initialize(rating)
    @rating = rating
    @next = nil
    @prev = nil
  end
end

class StackPointerCocktailSort
  attr_accessor :top, :size
  def initialize
    @top = nil
    @size = 0
  end
  def push(rating)
    node = Node.new(rating)
    if @top.nil?
      @top = node
    else
      node.prev = @top
      @top.next = node
      @top = node
    end
    @size += 1
  end
  def cocktail_sort
    return if @size <= 1
    base = @top
    while base&.prev
      base = base.prev
    end
    swapped = true
    start = base
    end_node = @top
    while swapped
      swapped = false
      cur = start
      while cur != end_node
        if cur.rating > cur.next.rating
          cur.rating, cur.next.rating = cur.next.rating, cur.rating
          swapped = true
        end
        cur = cur.next
      end
      break unless swapped
      swapped = false
      end_node = end_node.prev
      cur = end_node
      while cur != start
        if cur.prev.rating > cur.rating
          cur.rating, cur.prev.rating = cur.prev.rating, cur.rating
          swapped = true
        end
        cur = cur.prev
      end
      start = start.next
    end
  end
end

if ARGV.size < 2
  puts "Uso: ruby stack_pointer_cocktail_sort.rb <arquivo.csv> <N>"
  exit
end

path = ARGV[0]
n = ARGV[1].to_i
pilha = StackPointerCocktailSort.new

File.open(path) do |f|
  f.readline
  f.each_line.take(n).each do |line|
    user_id, movie_id, rating, timestamp = line.strip.split(',')
    pilha.push(Rating.new(user_id.to_i, movie_id.to_i, rating.to_f, timestamp.to_i))
  end
end

mem_before = ObjectSpace.memsize_of_all
time = Benchmark.realtime { pilha.cocktail_sort }
mem_after = ObjectSpace.memsize_of_all

puts "Ordenado #{n} registros em #{(time * 1000).round(2)} ms"
puts "Memória usada: #{((mem_after - mem_before) / 1024.0).round(2)} KB"
puts "Primeiros 100 registros ordenados:"
cur = pilha.top
while cur&.prev
  cur = cur.prev
end
100.times do
  break if cur.nil?
  puts cur.rating
  cur = cur.next
end

puts "Últimos 100 registros ordenados:"
cur = pilha.top
cur = cur.next while cur&.next
100.times do
  break if cur.nil?
  puts cur.rating
  cur = cur.prev
end